### PR TITLE
fix(backend-core): New Resource class structure

### PR DESCRIPTION
**Pull request by igneel64**
## Type of change

- [x] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/clerk-expo`
- [x] `@clerk/backend-core`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/edge`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

In general this PR addresses the creation of Resource classes on the backend-core side of the apps. It makes it more straightforward and removes some unneeded structures/checks.

If this structure is accepted, it will be applied to the rest of the API classes.

### Changes
- Readonly attributes of all API classes.
- Removal of unneeded 'Association` structure.
- Different initialization signature.
- Correct types based on the existence of attributes.
- Remove `camelcase-keys` as no longer needed

### Note
If we need to provide default values instead of `null`s on fields that are expected to be strings, we can also add that if it seems more intuitive.

### CODE REVIEW

**Comment by igneel64** [comment]
This is the example response object from the new and old versions, which is identical.

image

The input I would like is if we should consider this change breaking. The clerk-sdk-node re-exports these classes but we have no indication how they are to be used and actually they don't even have any methods except `fromJSON` attached to them. This is used only internally. WDYT ?

File: packages/backend-core/src/api/resources/User.ts [hunk 1]
```diff
@@ -1,84 +1,79 @@
...
-  emailAddresses: EmailAddress[];
-  phoneNumbers: PhoneNumber[];
-  web3Wallets: Web3Wallet[];
-  externalAccounts: ExternalAccount[];
-}
-
-interface UserPayload extends UserProps, UserAssociations {}
-
-export interface User extends UserPayload {}
+export const userAssociations = {
```
Comment (igneel64) [inline packages/backend-core/src/api/resources/User.ts:11]: This is left here only until we decide if we want to keep this structure or not.

**Review by npetridis — APPROVED** [review]

File: packages/backend-core/src/api/resources/User.ts (hunk 1 above)
Comment (SokratisVidros) [reply packages/backend-core/src/api/resources/User.ts:11]: A simpler modelling with just arrays would be better but this should be part of future work

**Review by SokratisVidros — COMMENTED** [review]
@igneel64 This looks fine for the User resource. We should do the other resources as well.

File: packages/backend-core/src/api/resources/User.ts [hunk 2]
```diff
@@ -1,84 +1,79 @@
...
-    Object.assign(obj, filtered);
+    const obj = camelcaseKeys(data) as Record<string, any>;
 
     obj.emailAddresses = (data.email_addresses || []).map(x => EmailAddress.fromJSON(x));
     obj.phoneNumbers = (data.phone_numbers || []).map(x => PhoneNumber.fromJSON(x));
     obj.web3Wallets = (data.web3_wallets || []).map(x => Web3Wallet.fromJSON(x));
     obj.externalAccounts = (data.external_accounts || []).map((x: ExternalAccountJSON) => ExternalAccount.fromJSON(x));
 
-    return new User(obj as UserPayload);
+    return new User(
```
[the remark this answers is not in the document]
Comment (igneel64) [reply packages/backend-core/src/api/resources/User.ts:53]: We do not, this is internal as stated.
image

In the update I will be changing positionals to a UserPayload type with `readonly` attributes as it is done now.

[the remark this answers is not in the document]
Comment (igneel64) [reply packages/backend-core/src/api/resources/User.ts:53]: Btw this seems **like** a factory already.

[the remark this answers is not in the document]
Comment (nikosdouvlis) [reply packages/backend-core/src/api/resources/User.ts:53]: @igneel64 Thanks for pointing out that this is the `fromJSON` method instead. The very first point about positions args still makes sense but please ignore the rest of my comment, I though I was looking at the body of the constructor :)

File: packages/backend-core/package.json [hunk 3]
```diff
@@ -13,7 +13,6 @@
   "module": "dist/mjs/index.js",
   "dependencies": {
     "@clerk/types": "^2.9.0",
-    "camelcase-keys": "^7.0.1",
```
Comment (nikosdouvlis) [inline packages/backend-core/package.json:16]: 🎈

**Review by nikosdouvlis — APPROVED** [review]
I slightly prefer named vs positional arguments, but since the ctor use is abstracted by the `fromJSON` method I'm good with it. This looks great 💯

**Review by SokratisVidros — APPROVED** [review]

**Pull request merged**